src/librustc_mir_build/hair/pattern/_match.rs

```diff
@@ -588,12 +588,11 @@ impl<'a, 'tcx> MatchCheckCtxt<'a, 'tcx> {
     crate fn create_and_enter<R>(
         tcx: TyCtxt<'tcx>,
         param_env: ty::ParamEnv<'tcx>,
+        pattern_arena: &'a TypedArena<Pat<'tcx>>,
         module: DefId,
-        f: impl FnOnce(MatchCheckCtxt<'_, 'tcx>) -> R,
+        f: impl FnOnce(MatchCheckCtxt<'a, 'tcx>) -> R,
     ) -> R {
-        let pattern_arena = TypedArena::default();
-
-        f(MatchCheckCtxt { tcx, param_env, module, pattern_arena: &pattern_arena })
+        f(MatchCheckCtxt { tcx, param_env, module, pattern_arena })
```

Comment:
The whole point of `create_and_enter` was to rank-2 encapsulate the arena, wasn't it ? I think you could replace this with a `new` function that returns a `MatchCheckCtxt`. Or even just create the `MatchCheckCtxt` directly since the function does literally nothing.

Reply:
I have only the vaguest idea of what "rank-2 encapsulate" means, but yes, `create_and_enter` isn't necessary :)  I have added extra commits to remove it, thanks for the suggestion.

Reply:
Yeah that term might have been unnecessarily show-off-y on my part sorry ><